doc/languages-frameworks/nim.section.md

```diff
@@ -1,82 +1,73 @@
 # Nim {#nim}
 
-## Overview {#nim-overview}
-
-The Nim compiler, a builder function, and some packaged libraries are available
-in Nixpkgs. Until now each compiler release has been effectively backwards
-compatible so only the latest version is available.
-
-## Nim program packages in Nixpkgs {#nim-program-packages-in-nixpkgs}
-
-Nim programs can be built using `nimPackages.buildNimPackage`. In the
-case of packages not containing exported library code the attribute
-`nimBinOnly` should be set to `true`.
+The Nim compiler and a builder function is available.
+Nim programs are built using `buildNimPackage` and a lockfile containing Nim dependencies.
 
 The following example shows a Nim program that depends only on Nim libraries:
-
 ```nix
-{ lib, nimPackages, fetchFromGitHub }:
+{ lib, buildNimPackage, fetchFromGitHub }:
 
-nimPackages.buildNimPackage (finalAttrs: {
+buildNimPackage (finalAttrs: {
   pname = "ttop";
-  version = "1.0.1";
-  nimBinOnly = true;
+  version = "1.2.7";
 
   src = fetchFromGitHub {
     owner = "inv2004";
     repo = "ttop";
     rev = "v${finalAttrs.version}";
-    hash = "sha256-x4Uczksh6p3XX/IMrOFtBxIleVHdAPX9e8n32VAUTC4=";
+    hash = "sha256-oPdaUqh6eN1X5kAYVvevOndkB/xnQng9QVLX9bu5P5E=";
   };
 
-  buildInputs = with nimPackages; [ asciigraph illwill parsetoml zippy ];
+  lockFile = ./lock.json;
 
+  nimFlags = [
+    "-d:NimblePkgVersion=${finalAttrs.version}"
+  ];
 })
 ```
 
-## Nim library packages in Nixpkgs {#nim-library-packages-in-nixpkgs}
-
+## Lockfiles {#nim-lockfiles}
+Nim lockfiles are created with the `nim_lk` utility.
+Run `nim_lk` with the source directory as an argument and it will print a lockfile to stdout.
+```sh
+$ cd nixpkgs
+$ nix run -f . nim_lk -- $(nix eval --raw -f . ttop.src) > pkgs/by-name/tt/ttop/lock.json
+```
 
-Nim libraries can also be built using `nimPackages.buildNimPackage`, but
-often the product of a fetcher is sufficient to satisfy a dependency.
-The `fetchgit`, `fetchFromGitHub`, and `fetchNimble` functions yield an
-output that can be discovered during the `configurePhase` of `buildNimPackage`.
+## Lockfile dependency overrides {#nim-lockfile-overrides}
```

Comment:
Is it possible for downstream users to add their own overrides, maybe on a package level? Only supporting one global override file is pretty limiting

Reply:
This is a problem, and so is not being able to override the `buildInputs` supplied by an override.

Reply:
I made the overrides overriddable by make a `nimOverrides` argument for `callPackage` that is passed to `buildNimPackage`.

Reply:
Could be something like this, haven't tested
```nix
updateScript = writeShellScript "update-${finalAttrs.finalPackage.pname}" {
  nativeBuildInputs = [ nim_lk nix-update ];
} ''
  nix-update "$UPDATE_NIX_ATTR_PATH"
  src="$(nix build -f . "$UPDATE_NIX_ATTR_PATH".src --no-link --print-out-paths)"
  lockFilePath="''${
    nix eval --impure  --expr 'with import ./. {}; lib.removePrefix "${toString ./.}/" '"$UPDATE_NIX_ATTR_PATH"'.lockFile' | cut -d: -f1
  }"
  nim_lk "$src" > "$lockFilePath"
''
```

Reply:
That would be less verbose in the typical case but I like making the overrides visible and explicit. I will think about it.

Reply:
`NIX_SHOW_STATS` shows an improvement when the overrides are included in `buildNimPackage`, so I've refactored to do that.